blockstorage: do not flush block to disk if it is already there

pinheadmz opened the pull request:
Closes https://github.com/bitcoin/bitcoin/issues/2039

When reindexing from flat-file block storage there is no need to write anything back to disk, since the block data is already there. This PR skips flushing to disk those blocks that already have a known position in the datastore. Skipping this means that users can write-protect the `blk` files on disk which may be useful for security or even safely sharing that data between multiple bitcoind instances.

`FindBlockPos()` may also flush the undo data file, but again this is skipped if the corresponding block position is known, like during the initial stage of a reindex when block data is being indexed. Once the block index is complete the validation mechanism will call `ConnectBlock()` which will save undo data at that time.

The call stack looks like this:

```
init()
ThreadImport() <-- process fReindex flag
LoadExternalBlockFile()
AcceptBlock()
SaveBlockToDisk()
FindBlockPos()
FlushBlockFile() <-- unnecessary if block is already on disk
```

A larger refactor of this part of the code was started by mzumsande here:  https://github.com/mzumsande/bitcoin/tree/202207_refactor_findblockpos including this fix, reviewers can let me know if the changes should be combined.

brunoerg commented on test/functional/feature_reindex.py:
```diff
@@ -68,12 +73,32 @@ def find_block(b, start):
+        self.stop_node(1)
+
+        # make the first block file read-only
+        path = os.path.join(self.nodes[1].chain_path, 'blocks', 'blk00000.dat');
```
```suggestion
        path = os.path.join(self.nodes[1].chain_path, 'blocks', 'blk00000.dat')
```

pinheadmz replied:
thanks! fixed.

brunoerg commented on test/functional/feature_reindex.py:
```diff
@@ -19,22 +19,27 @@
-        self.stop_nodes()
-        extra_args = [["-reindex-chainstate" if justchainstate else "-reindex"]]
-        self.start_nodes(extra_args)
+        self.stop_node(0)
```
Instead of stopping the node and right after starting it again, you could use `self.restart_node`.

pinheadmz replied:
ah thanks, fixed. also added an explanation of the test to the comments at the top

LarryRuane commented on test/functional/feature_reindex.py:
```diff
@@ -68,12 +72,32 @@ def find_block(b, start):
+        self.connect_nodes(0, 1)
+        # generate enough blocks to ensure that the -fastprune node fills up the
+        # first blk00000.dat file and starts another block file
+        self.generatetoaddress(self.nodes[1], 300, self.nodes[1].get_deterministic_priv_key().address)
```
Verified:
```
$ ls -l /tmp/btest/node1/regtest/blocks/blk*.dat
-rw------- 1 larry larry 65400 Feb  3 16:58 /tmp/btest/node1/regtest/blocks/blk00000.dat
-rw------- 1 larry larry 16384 Feb  3 16:58 /tmp/btest/node1/regtest/blocks/blk00001.dat
```

brunoerg replied:
To ensure it, I think you could check if both files exist:
```diff
diff --git a/test/functional/feature_reindex.py b/test/functional/feature_reindex.py
index c68773e06..a3c9562f1 100755
--- a/test/functional/feature_reindex.py
+++ b/test/functional/feature_reindex.py
@@ -84,6 +84,9 @@ class ReindexTest(BitcoinTestFramework):
         self.generatetoaddress(self.nodes[1], 300, self.nodes[1].get_deterministic_priv_key().address)
         self.stop_node(1)
 
+        assert os.path.exists(self.nodes[1].chain_path / 'blocks' / 'blk00000.dat')
+        assert os.path.exists(self.nodes[1].chain_path / 'blocks' / 'blk00001.dat')
+
         # make the first block file read-only
         path = os.path.join(self.nodes[1].chain_path, 'blocks', 'blk00000.dat')
         os.chmod(path, 0o444)
```

pinheadmz replied:
thanks, added.

LarryRuane commented on test/functional/feature_reindex.py:
```diff
@@ -68,12 +72,32 @@ def find_block(b, start):
+
+        # make the first block file read-only
+        path = os.path.join(self.nodes[1].chain_path, 'blocks', 'blk00000.dat')
+        os.chmod(path, 0o444)
```
Verified:
```
$ ls -l /tmp/btest/node1/regtest/blocks/blk*.dat
-r--r--r-- 1 larry larry 65400 Feb  3 16:58 /tmp/btest/node1/regtest/blocks/blk00000.dat
-rw------- 1 larry larry 16384 Feb  3 16:58 /tmp/btest/node1/regtest/blocks/blk00001.dat
```

LarryRuane commented on test/functional/feature_reindex.py:
```diff
@@ -68,12 +72,32 @@ def find_block(b, start):
+        os.chmod(path, 0o444)
+
+        # restart and reindex the node with the read-only block file
+        self.start_node(1, ["-reindex", "-fastprune"])
```
Verified that without the PR, the node doesn't start, and we see this in `debug.log`:
```
2023-02-04T00:05:32.219184Z [loadblk] [node/blockstorage.cpp:865] [ThreadImport] Reindexing block file blk00001.dat...
2023-02-04T00:05:32.224307Z [loadblk] [flatfile.cpp:44] [Open] Unable to open file /tmp/btest/node1/regtest/blocks/blk00000.dat
2023-02-04T00:05:32.224320Z [loadblk] [util/system.h:50] [error] ERROR: Flush: failed to open file 0
2023-02-04T00:05:32.224334Z [loadblk] [shutdown.cpp:26] [AbortNode] *** Flushing block file to disk failed. This is likely the result of an I/O error.
```
Nothing the PR needs to address, but just FYI, the first message may look confusing; it mentions `blk00001.dat` just before it attempts to write the previous file, `blk00001.dat`, which is where it runs into the write failure. Notice that this write is what the PR prevents because it's unnecessary. This is a nice way to test that the write happens without the PR, and doesn't happen with the PR.

LarryRuane commented on test/functional/feature_reindex.py:
```diff
@@ -68,12 +72,32 @@ def find_block(b, start):
+        assert_equal(self.nodes[1].getblockcount(), 312)
+
+        # shut down this node, triggering disk flushes, but not to the read-only file
+        with self.nodes[1].assert_debug_log(expected_msgs=['Flush'], unexpected_msgs=['Error']):
+            self.stop_node(1)
```
There are many `Flush` log lines for unrelated things (like the ChainState), so consider making that string more specific. Maybe:
```
2023-02-04T00:15:05.877123Z [shutoff] [logging/timer.h:57] [Log] [bench] FlushStateToDisk: write block and undo data to disk completed (6.62ms)
```
Similarly, maybe "Error" is too general? It does work, but I'd be afraid that something unrelated might log "Error" in the future and break this test. Maybe look for one of the error strings in my comment above?

pinheadmz replied:
good call, I was being too general, updated. thanks!

LarryRuane reviewed (commented):
Approach and code review ACK a6b05f8db63bae7eefb5330048d8b1441f1a9e5e
I don't understand why the test needs to start a second node or why `-fastprune` is needed (but that may just be me!).

EDIT: never mind, I see now why it's needed, you've documented it well.

> reviewers can let me know if the changes should be combined

It seems fine to me that this is a separate PR, but I don't have a strong opinion.

LarryRuane commented on test/functional/feature_reindex.py:
```diff
@@ -8,6 +8,7 @@
 - Stop the node and restart it with -reindex. Verify that the node has reindexed up to block 3.
 - Stop the node and restart it with -reindex-chainstate. Verify that the node has reindexed up to block 3.
 - Verify that out-of-order blocks are correctly processed, see LoadExternalBlockFile()
+- Start a second node, generate blocks, then restart with -reindex after setting blk files to read-only
```
If you need to retouch, maybe expand this comment to explain why two nodes are needed.

pinheadmz replied:
I think I added a second node to the test so I could start a clean chain using smaller `blk` files (`-fastprune`) without bugging the existing tests that switch a few options on and off throughout.

pinheadmz replied:
added comment to top of the test where second node is set up

LarryRuane reviewed (commented):
crtACK 386faaf76ea0edb2150725c5bdb69d234a2f2607
I'm still not able to figure out why two nodes are needed in the test. I tried modifying the test to use just one node, but the new test `reindex_readonly()`, didn't pass (the log file contained the unexpected string "failed to open file"). So the second node does seem to be needed.

LarryRuane reviewed (commented):
ACK ed57565e36bf514fc4d063972464c9152a08fb8e

jonatack reviewed (commented):
Concept ACK

pinheadmz commented on test/functional/feature_reindex.py:
```diff
@@ -68,12 +72,35 @@ def find_block(b, start):
+        self.connect_nodes(0, 1)
+        # generate enough blocks to ensure that the -fastprune node fills up the
+        # first blk00000.dat file and starts another block file
+        self.generatetoaddress(self.nodes[1], 300, self.nodes[1].get_deterministic_priv_key().address)
```
Thanks Jon I believe I addressed all the style nits and reduced the number of generated blocks as requested in ec7dab684f172c7df37b3a9fe16250bedc2b775f. Adding test coverage for flush to disk is next...

LarryRuane commented on test/functional/feature_reindex.py:
```diff
@@ -68,12 +72,49 @@ def find_block(b, start):
+        self.generatetoaddress(self.nodes[1], blocks_needed, addr)
+        self.stop_node(1)
+
+        assert os.path.exists(self.nodes[1].chain_path / 'blocks' / 'blk00000.dat')
+        assert os.path.exists(self.nodes[1].chain_path / 'blocks' / 'blk00001.dat')
```
Some assertions are not checking that the PR's changes are working correctly, they're ensuring that the test is setting things up as expected, and sometimes that's not obvious to readers. Sorry I didn't think to mention this earlier, but if you retouch, consider adding a comment before these two lines to make it clear that all we're doing here is verifying the test setup.

pinheadmz replied:
Ok good feedback. Yeah these asserts are more like sanity checks. Testing the test! Although they also fail early on certain code regressions for example if the `-fastprune` size is changed in the future.

LarryRuane commented on test/functional/feature_reindex.py:
```diff
@@ -68,12 +72,49 @@ def find_block(b, start):
+        block_count = self.nodes[1].getblockcount()
+        fastprune_blockfile_size = 0x10000
+        size_needed = fastprune_blockfile_size - (block_size * block_count)
+        blocks_needed = size_needed // block_size
```
I like the way you're calculating the number of blocks needed rather than hard-coding the number; I verified that this value (when I run the test) is 240, which matches what @jonatack suggested.

LarryRuane reviewed (commented):
ACK e7a0339b8cb612088ca212c568c0499c0f02398b

pinheadmz commented:
Rebased on master and prepended an extra commit with a unit test for BlockManager. The test passes on master and the branch without modification, hopefully illustrating that at least some behaviors are not affected.

It's hard to test precisely for the branch's change in a unit test because file-flushing is not something we can really predict or test for (sometimes it happens as soon as we `fwrite()`). 

The actual error in issue #2039 comes from an attempt to OPEN a file with `rw` when the file is read-only:

https://github.com/bitcoin/bitcoin/blob/82793f1984911774b111117f2e81d5f3b0bbec68/src/flatfile.cpp#L81-L86

...and this PR avoids the call to `Flush()` altogether when we know it is not necessary

@jonatack lemme know if the new test is reassuring enough, or if you have advice for better coverage?

hebasto reviewed (commented):
Tested that this PR indeed fixes #2039:

- on master (82793f1984911774b111117f2e81d5f3b0bbec68):
```
$ chmod u-w ~/.bitcoin/signet/blocks/blk00003.dat
$ ./src/bitcoind -signet -reindex
...
2023-02-27T17:58:31Z Reindexing block file blk00003.dat...
2023-02-27T17:58:31Z Unable to open file /home/hebasto/.bitcoin/signet/blocks/blk00003.dat
2023-02-27T17:58:31Z ERROR: Flush: failed to open file 3
2023-02-27T17:58:31Z *** Flushing block file to disk failed. This is likely the result of an I/O error.
2023-02-27T17:58:31Z Error: A fatal internal error occurred, see debug.log for details
Error: A fatal internal error occurred, see debug.log for details
...
```

- this PR (5afd6bf274cf50743dc345eb9c0520c1b64c8069)
```
$ chmod u-w ~/.bitcoin/signet/blocks/blk00003.dat
$ ./src/bitcoind -signet -reindex  # no errors
```

mzumsande reviewed (commented):
Concept ACK, will review more closely soon.

The added unit test seems to fail under Win64.

pinheadmz commented:
force-push to e600eb442802ca7b23189c8ceee448730e22e75a:
- rebased on master and fixed conflict
- added 2-second sleep before calls to `SaveBlockToDisk()`. This is to ensure that if the file is modified, its modification timestamp will definitely be updated. This test was intermittent on Windows I think because FAT has a 2-second resolution for `mtime` values. 🤷‍♂️

mzumsande reviewed (commented):
Tested ACK e600eb442802ca7b23189c8ceee448730e22e75a

I feel like one small step is missing in the motivation why this is safe: `FlushBlockFile()` not only flushes the blk file, but possibly also the undo file (and does this no longer during reindex after this change), so it would be good to mention somewhere (e.g. in the OP) why this is ok.
My understanding is that new undo data will only be available after connecting blocks to the chain, which only happens in the second part of the reindex where we rebuild the chain - so flushing in the first part would be a no-op anyway, and skipping it is ok.

Also, if you retouch, flake8 and Clang-format show a few minor nits for me in the added code.

pinheadmz commented on src/test/blockmanager_tests.cpp:
```diff
@@ -82,4 +83,105 @@ BOOST_FIXTURE_TEST_CASE(blockmanager_scan_unlink_already_pruned_files, TestChain
+
+    // Ensure that if the file is modified its timestamp is updated,
+    // even on Windows (FAT has 2 second resolution)
+    std::this_thread::sleep_for(std::chrono::milliseconds(2000));
```
Thanks, I guarded the sleep for WIN32, cleaned up the lint and mentioned the undo data in OP.

LarryRuane commented on src/test/blockmanager_tests.cpp:
```diff
@@ -82,4 +83,113 @@ BOOST_FIXTURE_TEST_CASE(blockmanager_scan_unlink_already_pruned_files, TestChain
+    node::ReadBlockFromDisk(read_block, pos2, params->GetConsensus());
+    BOOST_CHECK_EQUAL(read_block.nVersion, 2);
+
+    // blk00000.dat was not modified
+    const fs::file_time_type time2 = fs::last_write_time(file_path);
+    BOOST_CHECK(time1 == time2);
```
Well, on non-windows, it _may_ have been modified, but within the same second. This is kind of a minor point (probably don't need to change anything), but if in the future there's a bug such that `blk00000.dat` was modified, this test might not catch it. I verified this by hacking the test such that the file _is_ modified, and about once in 10 or 20 runs, the timestamp didn't change. (This is on Ubuntu.)

But I guess that's okay, because if there is such a bug, it will likely be caught quickly. It's just that I wish this test could be 100 percent solid. The only way I could think of to do that is to read and checksum the file, and compare the checksums instead of the timestamps. (Then you wouldn't need that sleep for WIN32.) It wouldn't be a performance problem, because block files are very small in this test.

Just something to consider, it may not be worth the trouble.

achow101 replied:
> The only way I could think of to do that is to read and checksum the file, and compare the checksums instead of the timestamps. 

Isn't the point that the file contents aren't changing? So checksum wouldn't work because the files will be the same whether it's rewritten or not. The only things that can be different are the filesystem's metadata on the modification time.

pinheadmz replied:
> The only things that can be different are the filesystem's metadata on the modification time, and these are not covered by a checksum.

It might be the same usefulness actually, I seem to remember that if file contents aren't changed, when you call `fwrite()` or `fflush()`, the modification timestamp won't be updated. I researched different ways to detect a `fflush()` and this unit test was the closest. Most systems flush when `fwrite()` is called, and then a subsequent `fflush()` doesn't do anything at all. The best I could do with this unit test was cover current behavior to ensure that the PR didn't break it.

So actually it might be reasonable to just ditch the timestamp stuff altogether and just ensure that file contents aren't modified.

The better test coverage is in the functional test which will throw on master when bitcoin tries to **open** the read-only file.

LarryRuane commented on src/test/blockmanager_tests.cpp:
```diff
@@ -82,4 +83,113 @@ BOOST_FIXTURE_TEST_CASE(blockmanager_scan_unlink_already_pruned_files, TestChain
+    const fs::file_time_type time3 = fs::last_write_time(file_path);
+
+    // Sanity check: blk00000.dat should have been modified writing the dummy blocks.
+    BOOST_CHECK(time2 != time3);
```
This seems more problematic than what I mentioned above because this can cause false-positive. Is the resolution one second? Could all that happened since `time2` was sampled be in the same second? But I ran the test many times and did not see the problem.

pinheadmz replied:
What I learned is that windows/FAT systems have 2-second resolution but Linux/EXT4 etc have ms or ns resolution. I had a hard time determining the best method for testing when a file is flushed but I think you're right, using timestamps is just asking for intermittency. I think I wanted to avoid hashing files because it means opening the file while blockmanager might already have it open... but I think that's going to be the best way to test.

pinheadmz replied:
ok trying another approach where we ALWAYS pause, to ensure incremental timestamps. But on windows we have to pause for 2 seconds. All other filesystems we only need to wait for 1 ms.

LarryRuane commented on src/node/blockstorage.cpp:
```diff
@@ -621,7 +621,7 @@ bool BlockManager::FindBlockPos(FlatFilePos& pos, unsigned int nAddSize, unsigne
     bool finalize_undo = false;
     if (!fKnown) {
-        while (m_blockfile_info[nFile].nSize + nAddSize >= (gArgs.GetBoolArg("-fastprune", false) ? 0x10000 /* 64kb */ : MAX_BLOCKFILE_SIZE)) {
+        while (m_blockfile_info[nFile].nSize + nAddSize >= (gArgs.GetBoolArg("-fastprune", false) ? 0x10000 /* 64 KiB */ : MAX_BLOCKFILE_SIZE)) {
```
as long as you're touching this line
```suggestion
        const unsigned int max_size{gArgs.GetBoolArg("-fastprune", false) ? 0x10000 /* 64 KiB */ : MAX_BLOCKFILE_SIZE};
        while (m_blockfile_info[nFile].nSize + nAddSize >= max_size) {
```

pinheadmz replied:
I'm going to revert this so it can be fixed in https://github.com/bitcoin/bitcoin/pull/27191

LarryRuane reviewed (commented):
Verified `clang-format-diff`. The unit test is nice! You may want to mention in the commit comment that it doesn't test this PR (I was a little confused at first because I was expecting that it was).

pinheadmz commented:
force-push to b4ce309ef5f2bed5ed1f861305317a0dca3dc5f4:

- addressed all nits and review comments by @LarryRuane 
- added 1ms delay in blockmanager test for all non-windows systems (cc: @mzumsande )

LarryRuane reviewed (commented):
ACK b4ce309ef5f2bed5ed1f861305317a0dca3dc5f4
If you retouch, consider the suggestion for the unit test commit comment here: https://github.com/bitcoin/bitcoin/pull/27039#pullrequestreview-1320666918 (I think you may have just missed it), but it's not important.

pinheadmz commented:
> consider the suggestion for the unit test commit comment here: [#27039 (review)](https://github.com/bitcoin/bitcoin/pull/27039#pullrequestreview-1320666918) (I think you may have just missed it), but it's not important.

ah doh! I added a sentence to the PR description but if I rebase again I'll add to the actual commit

mzumsande commented on src/test/blockmanager_tests.cpp:
```diff
@@ -82,4 +83,116 @@ BOOST_FIXTURE_TEST_CASE(blockmanager_scan_unlink_already_pruned_files, TestChain
+    std::this_thread::sleep_for(pause);
+
+    // Attempt, but fail, to save block 3 to original position of block 2
+    blockman.SaveBlockToDisk(block3, 3, chain, *params, &pos2);
```
I don't understand what this part of the test is testing - calling `SaveBlockToDisk` for a block at a specified but incorrect position is something that should never be possible in an actual, non-corrupted node unless I'm missing something. 
But if it somehow happened anyway, even if the existing block isn't overwritten, I'd imagine that it would still cause some havoc, like changing block file statistics [here](https://github.com/bitcoin/bitcoin/blob/c7f1d95f52883d7087b74f45f5e4ce1100d51149/src/node/blockstorage.cpp#L646-L648) for the added size of a block that is never actually written to disk, which would probably be bad.

pinheadmz replied:
1. What I'm asserting here is that if a position is passed in, the function called "save block to disk" does not save a block to disk. By using incorrect data I can easily prove that it is not actually written to disk.
2. The call to `AddBlock` you highlight here confuses and concerns me a bit - since we know a `known` block won't be written, why are we updating the metadata at all?

mzumsande replied:
> 2\. The call to `AddBlock` you highlight here confuses and concerns me a bit - since we know a `known` block won't be written, why are we updating the metadata at all?

Because the block index LevelDB database (saved in `blocks/index`) is being wiped at the beginning of a reindex and needs to be rebuilt from scratch as we look at each block. Part of this database is meta info about each blockfile ([see bitcoin wiki](https://en.bitcoin.it/wiki/Bitcoin_Core_0.11_(ch_2):_Data_Storage#Block_index_.28leveldb.29)). I think that's the main reason why we have to call `SaveBlockToDisk` for each block during a reindex at all.

pinheadmz replied:
ok make sense thanks. do you think my first point justifies the method of the test or is that just confusing then and I should remove it?

mzumsande replied:
I think that both removing it and adding an explanatory comment would be ok. Maybe the test could also check that the block statistics (`CBlockFileInfo`) is not the same after the call (`nBlocks` should have increased)?

pinheadmz replied:
added in force-push to 5e579c6435145d84e43fb79224f2cb09b40ea047 and rebased on master

pinheadmz commented:
force push to: 93c70287a6434c6c665a211dc4dfbbd9c3db4083
- rebase on master
- fix failing unit test

mzumsande reviewed (commented):
Code Review ACK 93c70287a6434c6c665a211dc4dfbbd9c3db4083

turkycat commented on src/test/blockmanager_tests.cpp:
```diff
@@ -82,4 +83,122 @@ BOOST_FIXTURE_TEST_CASE(blockmanager_scan_unlink_already_pruned_files, TestChain
+    FlatFilePos pos1{blockman.SaveBlockToDisk(block1, /*nHeight=*/1, chain, *params, /*dbp=*/nullptr)};
+
+    // Write second block
+    FlatFilePos pos2 = blockman.SaveBlockToDisk(block2, 2, chain, *params, nullptr);
```
nit: use of assignment operator on line 126 while using initializer syntax for the same type on 123.

```suggestion
    FlatFilePos pos2{blockman.SaveBlockToDisk(block2, /*nHeight=*2, chain, *params, /*dbp=*/nullptr)};
```

pinheadmz replied:
Thanks good eye. If I retouch I'll fix this.

turkycat commented on src/test/blockmanager_tests.cpp:
```diff
@@ -82,4 +83,122 @@ BOOST_FIXTURE_TEST_CASE(blockmanager_scan_unlink_already_pruned_files, TestChain
+    const fs::path file_path = node::GetBlockPosFilename(pos2);
+    const fs::file_time_type time1 = fs::last_write_time(file_path);
+
+    std::this_thread::sleep_for(pause);
```
can you explain why you're sleeping _after_ reading `last_write_time`? If I understand correctly, based on the other comments in this thread, the sleep is necessary because the file modified time may not be updated immediately after the contents are written. This test does pass on my linux machine (which has a significantly shorter resolution time), yet I feel that we should be sleeping immediately after the potential file write does/does not occur before checking the file time and verifying it? This comment applies to the other uses of `sleep_for` in this test.

turkycat reviewed (commented):
code review and local test run ACK

470ef396

(edited by @DrahtBot due to GitHub bug https://github.com/MarcoFalke/DrahtBot/issues/31)

pablomartin4btc reviewed (commented):
Concept ACK. (went thru notes and other related PRs mentioned on the [Review club meeting preparation webpage](https://bitcoincore.reviews/27039) - Please add label when possible)

stickies-v reviewed (commented):
Approach ACK 93c70287a6434c6c665a211dc4dfbbd9c3db4083 (modulo tests)

Conceptually, this seems straightforward: we don't want to update block files during reindex, so avoiding the flushing altogether is desireable.

I think the only potential for unexpected behaviour change is if there are places other than `BlockManager::SaveBlockToDisk()` where block files are modified (and potentially not flushed). Looking at the callsites of `FlatFileSeq::Open()` and all of its parent callsites, it seems like the only place where we open a block file with `read_only == false` is in `BlockManager::SaveBlockToDisk()`. From that, I think the concept and approach make sense and seem safe.

When it comes to the tests, I'm not yet convinced. I think relying on `fs::last_write_time` is fragile, and I don't know if the test is useful enough to warrant this hacky approach. I understand this flushing behaviour is inherently difficult to test, but I'm not sure the current approach is worth the hassle?

pinheadmz commented:
Ok I updated the unit test: removed the time modified checks because all we need to check is if the content of the file has been changed, which it is NOT if `dbp != nullptr`. The read-only check in the functional test covers the rest.

pinheadmz commented:
> Maybe rebase to make CI less sad?

rebased. cheer up little guy!

pinheadmz commented:
> 🐙 This pull request conflicts with the target branch and [needs rebase](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#rebasing-changes).

done ✅

pablomartin4btc commented:
re-ACK 46cca252e30f54b06df011d268ef1b8a48b076e7.

fanquake commented:
> done 

We are likely going to defer mergeing this for the moment, as it conflicts with the next two kernel PRs, and assumeutxo. Up to you, if you'd like to keep it rebased after each merge, in the interim.

pinheadmz commented:
> If this is done I think we could merge everything except the last two commits in a separate PR, and start having better test coverage in master right away. 

Great plan, good idea thanks. Unit and functional tests opened in separate PR: https://github.com/bitcoin/bitcoin/pull/27850 I'll return to this after the tests and other conflicting PRs are merged or closed.

luke-jr commented on src/node/blockstorage.cpp:
```diff
@@ -648,8 +648,8 @@ bool BlockManager::FindBlockPos(FlatFilePos& pos, unsigned int nAddSize, unsigne
     if ((int)nFile != m_last_blockfile) {
         if (!fKnown) {
             LogPrint(BCLog::BLOCKSTORE, "Leaving block file %i: %s\n", m_last_blockfile, m_blockfile_info[m_last_blockfile].ToString());
+            FlushBlockFile(/*fFinalize=*/true, finalize_undo);
```
Won't this result in the undo files not getting flushed when expected?

mzumsande replied:
See https://github.com/bitcoin/bitcoin/pull/27039#pullrequestreview-1320217718 for my thoughts on this,  a similar explanation was added to the OP.

pinheadmz commented:
@ryanofsky separate tests PR is ready for review: https://github.com/bitcoin/bitcoin/pull/27850

pinheadmz committed:
blockstorage: do not flush block to disk if it is already there

test: ensure we can reindex from read-only block files now

pinheadmz commented:
Rebased on master now that #27850 is merged. This PR becomes very slim, one line of code changed and switching the test from expected fail to expected pass.

mzumsande commented:
Should squash the two commits so that there isn't a commit for which the CI fails.

pinheadmz commented:
> Should squash the two commits so that there isn't a commit for which the CI fails.

done thanks

furszy reviewed (commented):
Not blocking if others are happy with the current state.

I would prefer to split up `FindBlockPos()` prior to introducing any specific performance improvement. This method mixes two different functionalities with zero documentation and a misleading name. I think that we could move forward faster if the function wouldn't be that obscure.

mzumsande commented:
> I would prefer to split up `FindBlockPos()` prior to introducing any specific performance improvement.

I have an old branch for that (linked in the OP), was originally planning to PR it once this gets merged.

pinheadmz commented:
Rebased on master to fix conflicts and address test changes in the followup https://github.com/bitcoin/bitcoin/pull/28660

Might be good to get review from @TheCharlatan because this PR touches code just updated in https://github.com/bitcoin/bitcoin/pull/27866

mzumsande reviewed (commented):
ACK 8a1ab9aedb6b2ba8222a62d5513b9b0bc68324bd

Getting some chattr warnings, but these are independent of the changes here, so I'll report them in #28660.

furszy reviewed (commented):
Code review ACK 8a1ab9aed

pinheadmz commented:
Thanks for the reviews, @mzumsande @furszy @pablomartin4btc @LarryRuane I just rebased on master to clean up CI

furszy reviewed (commented):
Rebase diff ACK dfcef53.

Please try to not rebase the PR when it is not needed. A not related CI failure, or an spurious one, will not block a PR from getting it merged.

achow101 commented:
ACK dfcef536d0e6c40e98dce35ae7af6e3e4a2595cd

pinheadmz commented on test/functional/feature_reindex_readonly.py:
```diff
@@ -73,10 +74,10 @@ def reindex_readonly(self):
-                self.nodes[0].assert_start_raises_init_error(extra_args=['-reindex', '-fastprune'],
-                    expected_msg="Error: A fatal internal error occurred, see debug.log for details")
+            self.log.debug("Attempt to restart and reindex the node with the unwritable block file")
+            with self.nodes[0].wait_for_debug_log([b"Reindexing finished"]):
```
I'm not sure I understand. The previous code expected a crash so the test would know exactly when to check for the expected log message. In the new code bitcoind just runs fine so we have to poll the log for the success message.

pinheadmz replied:
OK I see now thanks. So I guess... `start_node` calls `wait_for_rpc_connection`. Is RPC available before reindexing is finished? If we can rely on that to prevent race conditions then this could just be `assert_debug_log`.

pinheadmz replied:
Got it, thanks. Worth opening a new PR?
